src/sage/cpython/cython_metaclass.h

```diff
@@ -52,6 +52,15 @@ static CYTHON_INLINE int Sage_PyType_Ready(PyTypeObject* t)
     if (r < 0)
         return r;
 
+#if PY_VERSION_HEX >= 0x03050000
+    // Cython 3 sets Py_TPFLAGS_HEAPSIZE before calling PyType_Ready,
+    // and resets just after the call. We need to do the same, or it may
+    // end up with illegal memory accesses.
```

Comment:
This `Sage_PyType_Ready` function is supposed to be used in place of `PyType_Ready` using macros:
```
#define PyType_Ready(t)  Sage_PyType_Ready(t)
```

As a result, when Cython calls `PyType_Ready` in their code, it is actually calling `Sage_PyType_Ready` here. As seen in the Cython issue, `Py_TPFLAGS_HEAPTYPE` should only be set for the duration of `PyType_Ready`, and all the following code in `Sage_PyType_Ready` should not run with it set.

Not having this results in segfaults for the first example in `src/sage/cpython/cython_metaclass.pyx`. In particular, the `print` line tries to print the string representation of `MyCustomType`, which have illegal memory accesses.
```
sage: cython(                                                               # optional - sage.misc.cython
....: '''
....: cimport cython
....: cimport sage.cpython.cython_metaclass
....: cdef class MyCustomType():
....:     @cython.always_allow_keywords(False)
....:     def __getmetaclass__(_):
....:         class MyMetaclass(type):
....:             def __init__(*args):
....:                 print("Calling MyMetaclass.__init__{}".format(args))                                                                                                                                                                                      ....:         return MyMetaclass
....:
....: cdef class MyDerivedType(MyCustomType):
....:     pass
....: ''')
```

Unfortunately, I think it would be unconvenient to add a test without copying the entire header file.

Reply:
I see. Maybe rephrase to something like "We need to reset it earlier, since otherwise it may end up with illegal memory accesses."

Note that in this comment you said `Py_TPFLAGS_HEAPSIZE` but you want to say `Py_TPFLAGS_HEAPTYPE` (I think).